src/ws.nim

```
@@ -0,0 +1,415 @@
+    ret.add(f.maskKey[0].uint8)
+    ret.add(f.maskKey[1].uint8)
+    ret.add(f.maskKey[2].uint8)
+    ret.add(f.maskKey[3].uint8)
+
+  # Write the data.
+  ret.add(data)
+  return ret
+
+proc send*(ws: WebSocket, data: seq[byte], opcode = Opcode.Text): Future[
+    void] {.async.} =
+  try:
+    var maskKey: array[4, char]
+    if ws.masked:
+      maskKey = genMaskKey(ws.rng)
+    var frame = encodeFrame((
+      fin: true,
+      rsv1: false,
+      rsv2: false,
+      rsv3: false,
+      opcode: opcode,
+      mask: ws.masked,
+      data: data,
+      maskKey: maskKey
+    ))
+    const maxSize = 1024*1024
+    # Send stuff in 1 megabyte chunks to prevent IOErrors.
+    # This really large packets.
+    var i = 0
+    while i < frame.len:
+      let frameSize = min(frame.len, i + maxSize)
+      let res = await ws.tcpSocket.write(frame[i ..< frameSize])
+      if res != frameSize:
+        raise newException(ValueError, "Error while send websocket frame")
+      i += maxSize
+  except IOError, OSError, ValueError:
+    # Wrap all exceptions in a WebSocketError so its easy to catch
+    raise newException(WebSocketError, "Failed to send data: " &
+        getCurrentExceptionMsg())
+
+proc sendStr*(ws: WebSocket, data: string, opcode = Opcode.Text): Future[
+                void] {.async.} =
+  await send(ws, toBytes(data), opcode)
+
+proc close*(ws: WebSocket) =
+  ## Close the Socket, sends close packet.
+  ws.readyState = Closed
+  proc close() {.async.} =
+    await ws.send(@[], Close)
+    ws.tcpSocket.close()
+  asyncCheck close()
+
+proc receiveFrame(ws: WebSocket): Future[Frame] {.async.} =
+  ## Gets a frame from the WebSocket.
+  ## See https://tools.ietf.org/html/rfc6455#section-5.2
+
+  # Grab the header.
+  var header = newSeq[byte](2)
+  try:
+    await ws.tcpSocket.readExactly(addr header[0], 2)
+  except TransportUseClosedError:
+    ws.readyState = Closed
+    raise newException(WebSocketError, "Socket closed")
+  except CatchableError:
```

By catching `CatchableError` you swallowing cancellation exceptions, so its better to avoid such construction or comment about reasons.

---

Done